design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,230 @@
+This means each workload must be updated to use a `pc-q35-rhel9.x.x` machine type.
+````
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
+	* if currently running, take VM offline immediately and update machine type
+	* if not currently running, update machine type immediately
+* Allow user to specify certain VMs for updating machine type automation (e.g. by namespace)
+
+## Non Goals
+While these additions could be beneficial to cluster-admins who wish to use the mass machine type transition (MMTT), these are not covered in the scope of this design proposal: 
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Allowing user to filter VMs in ways other than by namespace; e.g. with label-selector
+
+Also not covered in this design proposal is the method in which users/cluster admins will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with CentOS Stream 10 without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. Both the minimum supported machine type and the target machine type are constants that can be updated as future versions are released and more machine types are deprecated upon these releases. 
+
+If the VM is running, the update will not take effect until it is restarted, so the label `restart-vm-required` will be applied to the VM until the VMI is removed (indicating the VM has been stopped), upon which the label will be cleared. The job will only terminate once all `restart-vm-required` labels have been cleared. By default, the user must handle stopping/restarting the labelled VMs manually. This is to allow the user to safely update the machine types of their VMs at a time they choose without worrying about workload interruptions. However, the user also has the option to allow the job to automatically restart every running VM that is being updated and apply the changes immediately.
```

Comment:
I'm not sure if this should run inside a `Job` or if an "operator" would be more suited to do this? The job could potentially run for a very very long time and might even be stopped/killed before all VMs were updated?

Reply:
My opinion is that, in general, we already treat our components as monoliths more than micro-services. For example, virt-controller, just like `util.go`, is a generic name for "insert any logic you need here". This caused `virt-controller`, just like many of our other core components, to not be cohesive and deal with many unrelated tasks. IMHO this is a serious problem that hurts us in many ways, e.g. performance, scalability, maintainability, and will only get more severe over time if we don't change how things are being done.

IMO, This is a perfect example.

We're talking about a rare operation that an admin would likely to perform once in a few years. This operation is self-contained, has a very specific goal that is unrelated to virt-contoller's role, and I see no reason it should be handled by virt-controller or any other core component.

> The job could potentially run for a very very long time

The job would run as long as the cluster is not fully updated. As I see it, this is a feature, not a bug. BTW, by "running" I basically mean that the process would sleep, waiting for events about VMIs being updated, so in terms of performance I find it negligible. But, in terms of usability, I think that's actually great. As an admin I would know the status of the operation in a single and dedicated place.

Reply:
IMO it doesn't have to be in the `virt-controller`, but it could be in an independent micro-service too.

If it is really just a operation performed once in a few years then I'd agree with keeping it as simple as possible and using a Job instead of a long running operator.

As I said in my other comment, I was just concerned about the state of the Job and it maybe being killed before it was able to complete the operation. So an admin would have to monitor it until it eventually succeeded.